tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -12,90 +12,457 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
+
 // Provide helper routine for obtaining  gpu target information useful
 // for llvm IR contruction.
 
 #include "tensorflow/compiler/xla/service/gpu/target_util.h"
 
+#include "absl/strings/str_cat.h"
 #include "llvm/IR/MDBuilder.h"
+#include "tensorflow/compiler/xla/primitive_util.h"
+#include "tensorflow/compiler/xla/service/llvm_ir/ir_builder_mixin.h"
 #include "tensorflow/compiler/xla/service/llvm_ir/llvm_util.h"
+#include "tensorflow/compiler/xla/xla_data.pb.h"
 #include "tensorflow/core/platform/logging.h"
 
 namespace xla {
+
+using absl::StrAppend;
+
 namespace gpu {
 namespace {
+
 // Utility functions to obtain NVPTX/AMDGPU specific information.
 
-// Wrapper structure for carrying llvm intrinsic ids for NVPTX/AMDGPU platforms.
-struct TargetIntrinsics {
-  llvm::Intrinsic::ID nvptx_intrinsic;
-  llvm::Intrinsic::ID amdgpu_intrinsic;
+// Wrapper structure for  intrinsic information for NVPTX/AMDGPU.
+struct TargetIntrinsicInfo {
+  TargetIntrinsicInfo(llvm::Intrinsic::ID x) : intrinsic(x) {}
+  llvm::Intrinsic::ID intrinsic;
+};
+
+// Wrapper structure for  device functionc information for NVPTX/AMDGPU.
```

Comment:
This struct seems to have something to do with callsites, so this comment seems incomplete?